File: analytics/src/main/java/io/appmetrica/analytics/impl/utils/file/SuspendableFileLocker.java
package io.appmetrica.analytics.impl.utils.file;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.appmetrica.analytics.impl.FileProvider;
import io.appmetrica.analytics.impl.IOUtils;
import io.appmetrica.analytics.impl.Utils;
import java.io.File;
import java.io.RandomAccessFile;
import java.nio.channels.FileChannel;
import java.nio.channels.FileLock;
import java.util.HashMap;
import java.util.concurrent.Semaphore;

public final class SuspendableFileLocker {

    private static final HashMap<String, SuspendableFileLocker> sLocks = new HashMap<String, SuspendableFileLocker>();

    public synchronized static SuspendableFileLocker getLock(
            @NonNull Context context, @NonNull String simpleFileName
    ) {
        SuspendableFileLocker locker = sLocks.get(simpleFileName);
        if (locker == null) {
            locker = new SuspendableFileLocker(context, simpleFileName);
            sLocks.put(simpleFileName, locker);
        }
        return locker;
    }

    @NonNull
    private final String mLockFileName;

    private FileLock lock;
    private FileChannel channel;
    @Nullable
    private final File lockFile;
    private RandomAccessFile stream;
    private Semaphore semaphore = new Semaphore(1, true);

    private SuspendableFileLocker(@NonNull Context context, @NonNull String simpleFileName) {
        mLockFileName = simpleFileName + ".lock";
        FileProvider fileProvider = new FileProvider();
        File lockDir = fileProvider.getFileByPath(context.getCacheDir(), "appmetrica_locks");
        if (lockDir != null) {
            lockDir.mkdirs();
        }
        lockFile = fileProvider.getFileByPath(lockDir, mLockFileName);
    }

    public synchronized void lock() throws Throwable {
        semaphore.acquire();
        if (lockFile == null) {
            throw new IllegalStateException("Lock file is null");
        }
        if (channel == null) {
            stream = new RandomAccessFile(lockFile, "rw");
            channel = stream.getChannel();
        }
        lock = channel.lock();
    }

    public synchronized void unlock() {
        semaphore.release();
        if (semaphore.availablePermits() > 0) {
            IOUtils.releaseFileLock(mLockFileName, lock);
            Utils.closeCloseable(channel);
            Utils.closeCloseable(stream);
            channel = null;
            stream = null;
        }
    }
}

File: ndkcrashes/src/main/kotlin/io/appmetrica/analytics/ndkcrashes/impl/utils/SuspendableFileLocker.kt
package io.appmetrica.analytics.ndkcrashes.impl.utils

import android.content.Context
import java.io.Closeable
import java.io.File
import java.io.IOException
import java.io.RandomAccessFile
import java.nio.channels.FileChannel
import java.nio.channels.FileLock
import java.util.concurrent.Semaphore

class SuspendableFileLocker private constructor(context: Context, simpleFileName: String) {
    private class Lock(private val lockFile: File) {
        private val tag = "[SuspendableFileLocker]"

        private val stream: RandomAccessFile = RandomAccessFile(lockFile, "rw")
        private val channel: FileChannel = stream.channel
        private val lock: FileLock = channel.lock()

        fun release() {
            releaseFileLock(lockFile.name, lock)
            channel.safelyClose()
            stream.safelyClose()
        }

        private fun releaseFileLock(fileName: String?, lock: FileLock?) {
            if (lock != null && lock.isValid) {
                try {
                    lock.release()
                    DebugLogger.info(tag, "Lock released for $fileName.")
                } catch (e: IOException) {
                    DebugLogger.error(tag, "Failed to release lock for $fileName.", e)
                }
            }
        }

        private fun Closeable.safelyClose() = try {
            close()
        } catch (exception: Throwable) {
            // Do nothing
        }
    }

    private val semaphore = Semaphore(1, true)
    private val mLockFileName: String = "$simpleFileName.lock"
    private val lockFile: File? = context.cacheDir?.let { File(it, "appmetrica_locks") }?.let { lockDir ->
        lockDir.mkdirs()
        File(lockDir, mLockFileName)
    }

    private var lock: Lock? = null

    @Synchronized
    @Throws(Throwable::class)
    fun lock() {
        semaphore.acquire()
        checkNotNull(lockFile) { "Lock file is null" }
        lock = lock ?: Lock(lockFile)
    }

    @Synchronized
    fun unlock() {
        semaphore.release()
        if (semaphore.availablePermits() > 0) {
            lock?.release()
            lock = null
        }
    }

    companion object {
        private val sLocks = HashMap<String, SuspendableFileLocker>()

        @Synchronized
        fun getLock(context: Context, simpleFileName: String): SuspendableFileLocker {
            return sLocks.getOrPut(simpleFileName) { SuspendableFileLocker(context, simpleFileName) }
        }
    }
}
